src/nvim/buffer_defs.h

```diff
@@ -1204,6 +1204,9 @@ struct window_S {
   colnr_T w_skipcol;                // starting column when a single line
                                     // doesn't fit in the window
 
+  linenr_T w_last_topline;          ///< last known value for topline
```

Comment:
it would surprise me if these vars actually are needed. the condition for `win_viewport` should be more or less the same.

Reply:
Yep, I've changed it to use `w->w_viewport_invalid`.